// @wdio/selenium-standalone-service/build/launcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = __importDefault(require("@wdio/logger"));
const config_1 = require("@wdio/config");
const util_1 = require("util");
const fs_extra_1 = __importDefault(require("fs-extra"));
const SeleniumStandalone = __importStar(require("selenium-standalone"));
const utils_1 = require("./utils");
const DEFAULT_LOG_FILENAME = 'wdio-selenium-standalone.log';
const log = logger_1.default('@wdio/selenium-standalone-service');
const DEFAULT_CONNECTION = {
    protocol: 'http',
    hostname: 'localhost',
    port: 4444,
    path: '/wd/hub'
};
class SeleniumStandaloneLauncher {
    constructor(_options, _capabilities, _config) {
        this._options = _options;
        this._capabilities = _capabilities;
        this._config = _config;
        this.watchMode = false;
        this._stopProcess = () => {
            if (this.process) {
                log.info('shutting down all browsers');
                this.process.kill();
            }
        };
        this.skipSeleniumInstall = Boolean(this._options.skipSeleniumInstall);
        // simplified mode
        if (this.isSimplifiedMode(this._options)) {
            this.args = Object.entries(this._options.drivers).reduce((acc, [browserDriver, version]) => {
                if (typeof version === 'string') {
                    acc.drivers[browserDriver] = { version };
                }
                else if (version === true) {
                    acc.drivers[browserDriver] = {};
                }
                return acc;
            }, { drivers: {} });
            this.installArgs = { ...this.args };
        }
        else {
            this.args = this._options.args || {};
            this.installArgs = this._options.installArgs || {};
        }
    }
    async onPrepare(config) {
        this.watchMode = Boolean(config.watch);
        if (!this.skipSeleniumInstall) {
            const install = util_1.promisify(SeleniumStandalone.install);
            await install(this.installArgs);
        }
        /**
         * update capability connection options to connect
         * to standalone server
         */
        const capabilities = (Array.isArray(this._capabilities) ? this._capabilities : Object.values(this._capabilities));
        for (const capability of capabilities) {
            const cap = capability.capabilities || capability;
            const c = cap.alwaysMatch || cap;
            if (!config_1.isCloudCapability(c) && utils_1.hasCapsWithSupportedBrowser(c)) {
                Object.assign(c, DEFAULT_CONNECTION, { ...c });
            }
        }
        /**
         * start Selenium Standalone server
         */
        const start = util_1.promisify(SeleniumStandalone.start);
        this.process = await start(this.args);
        if (typeof this._config.outputDir === 'string') {
            this._redirectLogStream();
        }
        if (this.watchMode) {
            process.on('SIGINT', this._stopProcess);
            process.on('exit', this._stopProcess);
            process.on('uncaughtException', this._stopProcess);
        }
    }
    onComplete() {
        // selenium should not be killed in watch mode
        if (!this.watchMode) {
            this._stopProcess();
        }
    }
    _redirectLogStream() {
        var _a, _b;
        const logFile = utils_1.getFilePath(this._config.outputDir, DEFAULT_LOG_FILENAME);
        // ensure file & directory exists
        fs_extra_1.default.ensureFileSync(logFile);
        const logStream = fs_extra_1.default.createWriteStream(logFile, { flags: 'w' });
        (_a = this.process.stdout) === null || _a === void 0 ? void 0 : _a.pipe(logStream);
        (_b = this.process.stderr) === null || _b === void 0 ? void 0 : _b.pipe(logStream);
    }
    isSimplifiedMode(options) {
        return options.drivers && Object.keys(options.drivers).length > 0;
    }
}
exports.default = SeleniumStandaloneLauncher;
